### ISSUE TITLE
Cannot compile with rebar or rebar3

### ISSUE THREAD

**Issue by marcphilipp** [processone/xmpp#5]
It is currently not possible to compile the module using rebar or rebar3 when neither `../fast_xml/include` nor `deps/fast_xml/include` exist by coincidence.

There is some logic in ejabberd's rebar.config.script that replaces `deps/fast_xml/include` to `_build/default/lib/fast_xml/include` so it will work when compiled with ejabberd. However, third-party ejabberd modules would like to also use it. Therefore, `deps/fast_xml/include` should be replaced by `_build/default/lib/fast_xml/include` or the `rebar.config.script` should be updated to also resolve it accordingly.

```
$ rebar3 compile 
===> Fetching rebar3_hex ({pkg,<<"rebar3_hex">>,<<"3.1.0">>})
===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/rebar3_hex-3.1.0.tar
===> Compiling rebar3_hex
===> Verifying dependencies...
===> Fetching fast_xml ({pkg,<<"fast_xml">>,<<"1.1.18">>})
===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/fast_xml-1.1.18.tar
===> Fetching pc ({pkg,<<"pc">>,<<"1.4.0">>})
===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/pc-1.4.0.tar
===> Compiling pc
===> Fetching stringprep ({pkg,<<"stringprep">>,<<"1.0.7">>})
===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/stringprep-1.0.7.tar
===> Fetching p1_utils ({pkg,<<"p1_utils">>,<<"1.0.6">>})
===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/p1_utils-1.0.6.tar
===> Compiling p1_utils
===> Compiling fast_xml
===> Compiling /Users/marcp/Repositories/xmpp/_build/default/lib/fast_xml/c_src/fxml.c
===> Compiling /Users/marcp/Repositories/xmpp/_build/default/lib/fast_xml/c_src/fxml_stream.c
===> Linking /Users/marcp/Repositories/xmpp/_build/default/lib/fast_xml/priv/lib/fxml.so
===> Linking /Users/marcp/Repositories/xmpp/_build/default/lib/fast_xml/priv/lib/fxml_stream.so
===> Compiling stringprep
===> Compiling /Users/marcp/Repositories/xmpp/_build/default/lib/stringprep/c_src/stringprep.cpp
===> Linking /Users/marcp/Repositories/xmpp/_build/default/lib/stringprep/priv/lib/stringprep.so
===> Compiling xmpp
===> Compiling src/xmpp_util.erl failed
include/xmpp.hrl:29: can't find include file "fxml.hrl"
```

---

### Original Issue Description

I'm trying to compile a module that uses this new library. I've a very simple `rebar.config` file:

    {deps, [
        {xmpp, ".*", {git, "https://github.com/processone/xmpp", {tag, "1.1.4"}}}
    ]}.

When I try to run `rebar3 eunit` or `rebar3 compile` I get an error message:

    $ rebar3 compile
    ===> Verifying dependencies...
    ===> Fetching xmpp ({git,"https://github.com/processone/xmpp",
                                    {ref,"c010141487632db1666bcbf64f90f42c0fde50aa"}})
    ===> Fetching rebar3_hex ({pkg,<<"rebar3_hex">>,<<"3.1.0">>})
    ===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/rebar3_hex-3.1.0.tar
    ===> Compiling rebar3_hex
    ===> Fetching fast_xml ({pkg,<<"fast_xml">>,<<"1.1.18">>})
    ===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/fast_xml-1.1.18.tar
    ===> Fetching pc ({pkg,<<"pc">>,<<"1.4.0">>})
    ===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/pc-1.4.0.tar
    ===> Compiling pc
    ===> Fetching stringprep ({pkg,<<"stringprep">>,<<"1.0.7">>})
    ===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/stringprep-1.0.7.tar
    ===> Fetching p1_utils ({pkg,<<"p1_utils">>,<<"1.0.6">>})
    ===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/p1_utils-1.0.6.tar
    ===> Compiling p1_utils
    ===> Compiling fast_xml
    ===> Compiling /Users/marcp/Desktop/erlang-test/myapp/_build/default/lib/fast_xml/c_src/fxml.c
    ===> Compiling /Users/marcp/Desktop/erlang-test/myapp/_build/default/lib/fast_xml/c_src/fxml_stream.c
    ===> Linking /Users/marcp/Desktop/erlang-test/myapp/_build/default/lib/fast_xml/priv/lib/fxml.so
    ===> Linking /Users/marcp/Desktop/erlang-test/myapp/_build/default/lib/fast_xml/priv/lib/fxml_stream.so
    ===> Compiling stringprep
    ===> Compiling /Users/marcp/Desktop/erlang-test/myapp/_build/default/lib/stringprep/c_src/stringprep.cpp
    ===> Linking /Users/marcp/Desktop/erlang-test/myapp/_build/default/lib/stringprep/priv/lib/stringprep.so
    ===> Compiling xmpp
    ===> Compiling _build/default/lib/xmpp/src/xmpp_util.erl failed
    _build/default/lib/xmpp/include/xmpp.hrl:29: can't find include file "fxml.hrl"

However, when I compile this module separately, it works:

    $ cd _build/default/lib/xmpp
    $ rebar3 compile
    ===> Fetching rebar3_hex ({pkg,<<"rebar3_hex">>,<<"3.1.0">>})
    ===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/rebar3_hex-3.1.0.tar
    ===> Compiling rebar3_hex
    ===> Verifying dependencies...
    ===> Fetching fast_xml ({pkg,<<"fast_xml">>,<<"1.1.18">>})
    ===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/fast_xml-1.1.18.tar
    ===> Fetching pc ({pkg,<<"pc">>,<<"1.4.0">>})
    ===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/pc-1.4.0.tar
    ===> Compiling pc
    ===> Fetching stringprep ({pkg,<<"stringprep">>,<<"1.0.7">>})
    ===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/stringprep-1.0.7.tar
    ===> Fetching p1_utils ({pkg,<<"p1_utils">>,<<"1.0.6">>})
    ===> Downloaded package, caching at /Users/marcp/.cache/rebar3/hex/default/packages/p1_utils-1.0.6.tar
    ===> Skipping p1_utils (from {pkg,<<"p1_utils">>,<<"1.0.6">>,
                                         <<"EF0951DDF38E92B7E479AF4B8DC950DF76AF8C1030432EF68B7FD7AD17C436FE">>}) as an app of the same name has already been fetched
    ===> Compiling p1_utils
    ===> Compiling fast_xml
    ===> Compiling /Users/marcp/Desktop/erlang-test/myapp/_build/default/lib/xmpp/_build/default/lib/fast_xml/c_src/fxml.c
    ===> Compiling /Users/marcp/Desktop/erlang-test/myapp/_build/default/lib/xmpp/_build/default/lib/fast_xml/c_src/fxml_stream.c
    ===> Linking /Users/marcp/Desktop/erlang-test/myapp/_build/default/lib/xmpp/_build/default/lib/fast_xml/priv/lib/fxml.so
    ===> Linking /Users/marcp/Desktop/erlang-test/myapp/_build/default/lib/xmpp/_build/default/lib/fast_xml/priv/lib/fxml_stream.so
    ===> Compiling stringprep
    ===> Compiling /Users/marcp/Desktop/erlang-test/myapp/_build/default/lib/xmpp/_build/default/lib/stringprep/c_src/stringprep.cpp
    ===> Linking /Users/marcp/Desktop/erlang-test/myapp/_build/default/lib/xmpp/_build/default/lib/stringprep/priv/lib/stringprep.so
    ===> Compiling xmpp

I've also posted this question on https://www.rebar3.org/discuss/587d1b6673e9731b00889294. They told me the `rebar.config.script` is somehow responsible for this. Any help would be greatly appreciated.

**Comment by tsloughter** [comment]
I don't know that it is the the .script :), you asked if it was and I said it I would think so since it worked when you build from the root of `xmpp` instead of as a dependency. In which case it is often a config issue like that that causes the difference.

**Comment by zinid** [comment]
I'm not sure this works with rebar3, AFAIR nobody tries to compile it with it.
Also, the library at this point is designed to work with ejabberd only (unless we clean it up, this will take a few month after several releases of ejabberd, because we still don't know exactly what other files we should move here from ejabberd). I'm telling this because I don't know whether you're building it inside ejabberd or trying to "plug" into a separate project.

**Comment by marcphilipp** [comment]
I'm trying to build an ejabberd module and want to use rebar3 to compile my code and run eunit tests.

**Comment by marcphilipp** [comment]
That should be a supported use case, shouldn't it?

**Comment by zinid** [comment]
Yes, but I personally not a big expert in ejabberd building script, there is some magic in it: 
https://github.com/processone/ejabberd/blob/master/rebar.config#L82

**Comment by zinid** [comment]
I mean these lines are probably to solve the problem in ejabberd you're trying to cope with ;)

**Comment by marcphilipp** [comment]
It looks like "deps/*" gets resolved in rebar.config.script of ejabberd somehow:
https://github.com/processone/ejabberd/blob/master/rebar.config.script#L143-L157

**Comment by zinid** [comment]
Seems so

**Comment by marcphilipp** [comment]
I got it to work by adding the following lines to `rebar.config`:

```erlang
{overrides, [
    {add, xmpp, [{erl_opts, [{i, "_build/default/lib/fast_xml/include"}]}]}
]}.
```

**Comment by marcphilipp** [comment]
I've updated the issue description since I realized that I had `../fast_xml/include` on my local machine and that's why it worked when I just compiled the module.

**Comment by rutkas** [comment]
I have just got similar error when trying to update the ejabberd from portage on gentoo system that appeared today:

```
>>> Compiling source in /var/tmp/portage/net-im/ejabberd-17.01/work/ejabberd-17.01 ...
make -j2 'REBAR=rebar -v' src
rebar -v skip_deps=true compile
INFO:  Loading plugin deps_erl_opts from /var/tmp/portage/net-im/ejabberd-17.01/work/ejabberd-17.01/plugins/deps_erl_opts.erl
==> rel (compile)
==> ejabberd-17.01 (compile)
INFO:  sh info:
        cwd: "/var/tmp/portage/net-im/ejabberd-17.01/work/ejabberd-17.01"
        cmd: mv src/XmppAddr.hrl include
Compiled asn1/XmppAddr.asn1
Compiled src/eldap_filter_yecc.yrl
Compiled src/ejabberd_config.erl
Compiled src/gen_mod.erl
src/mod_muc_room.erl:54: can't find include file "xmpp.hrl"
src/mod_muc_room.erl:281: undefined macro 'NS_MUC_ADMIN'
src/mod_muc_room.erl:489: undefined macro 'NS_MUCSUB_NODES_MESSAGES'
src/mod_muc_room.erl:709: undefined macro 'NS_MUCSUB_NODES_PARTICIPANTS'
..... etc...
```

Any ideas ?

**Comment by rutkas** [comment]
I have found a temp solution for gentoo ejbberd update, the problem is that there are no deps for erlang xmpp and fast_xml libs and also they are missing in ejabberd src package:

```
emerge dev-erlang/xmpp
emerge dev-erlang/fast_xml 
```

patch rebar.config in sources to add pathes

.... 
{erl_opts, [nowarn_deprecated_function, 
            {i, "include"}, 
            {i, "deps/fast_xml/include"}, 
            {i, "deps/xmpp/include"}, 
            {i, "/usr/lib64/erlang/lib/xmpp-1.1.9/include"}, 
            {i, "/usr/lib64/erlang/lib/fast_xml-1.1.22/include"}, 
.... 

{eunit_compile_opts, [{i, "tools"}, 
                      {i, "include"}, 
                      {i, "deps/fast_xml/include"}, 
                      {i, "deps/xmpp/include"}, 
                      {i, "/usr/lib64/erlang/lib/xmpp-1.1.9/include"}, 
                      {i, "/usr/lib64/erlang/lib/fast_xml-1.1.22/include"}]}. 
.... 

and then compile/install them manually:

ebuild /usr/portage/net-im/ejabberd/ejabberd-17.01.ebuild compile 
ebuild /usr/portage/net-im/ejabberd/ejabberd-17.01.ebuild install 
ebuild /usr/portage/net-im/ejabberd/ejabberd-17.01.ebuild qmerge

**Comment by manuel-rubio** [comment]
This is because this PR https://github.com/processone/xmpp/pull/3 is still not merged.

**Comment by manuel-rubio** [comment]
I think this PR could be closed (it's the same as #3)

**Comment by pankajsoni19** [comment]
I am having the same issue.

**Comment by weiss** [comment]
Building with Rebar3 should now be fixed thanks to @nosnilmot (see #51).
